File: frontend-api/src/main/java/no/nav/melosys/tjenester/gui/dto/dokument/JournalpostInfoDto.java
package no.nav.melosys.tjenester.gui.dto.dokument;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

import no.nav.melosys.domain.arkiv.ArkivDokument;
import no.nav.melosys.domain.arkiv.Journalpost;
import no.nav.melosys.domain.arkiv.Journalposttype;
import no.nav.melosys.domain.arkiv.LogiskeVedlegg;
import no.nav.melosys.domain.kodeverk.Mottaksretning;

public class JournalpostInfoDto {
    public final String journalpostID;
    public final Instant mottattDato;
    public final Instant journalforingDato;
    public final Mottaksretning mottaksretning;
    public final String avsenderEllerMottaker;
    public final DokumentDto hoveddokument;
    public final List<DokumentDto> vedlegg;

    public JournalpostInfoDto(String journalpostID,
                              Instant mottattDato,
                              Instant journalforingDato,
                              Mottaksretning mottaksretning,
                              String avsenderEllerMottaker,
                              DokumentDto hoveddokument,
                              List<DokumentDto> vedlegg) {
        this.journalpostID = journalpostID;
        this.mottattDato = mottattDato;
        this.journalforingDato = journalforingDato;
        this.mottaksretning = mottaksretning;
        this.avsenderEllerMottaker = avsenderEllerMottaker;
        this.hoveddokument = hoveddokument;
        this.vedlegg = vedlegg;
    }

    public static JournalpostInfoDto av(Journalpost journalpost) {
        return new JournalpostInfoDto(journalpost.getJournalpostId(),
            journalpost.getForsendelseMottatt(),
            journalpost.getForsendelseJournalfoert(),
            av(journalpost.getJournalposttype()),
            journalpost.getKorrespondansepartNavn(),
            new DokumentDto(journalpost.getHoveddokument().getDokumentId(), journalpost.getHoveddokument().getTittel()),
            lagVedlegg(journalpost.getVedleggListe(), journalpost.getHoveddokument().getLogiskeVedlegg()));
    }

    private static List<DokumentDto> lagVedlegg(List<ArkivDokument> vedlegg, List<LogiskeVedlegg> interneVedlegg) {
        List<DokumentDto> vedleggListe = new ArrayList<>();
        vedlegg.forEach(v -> vedleggListe.add(new DokumentDto(v.getDokumentId(), v.getTittel())));
        interneVedlegg.forEach(v -> vedleggListe.add(new DokumentDto(v.getTittel())));
        return vedleggListe;
    }

    public Instant hentGjeldendeTidspunkt() {
        return this.mottattDato != null ? this.mottattDato : this.journalforingDato;
    }

    public static Mottaksretning av(Journalposttype journalposttype) {
        Mottaksretning retning;
        switch (journalposttype) {
            case INN: retning = Mottaksretning.INN;
                break;
            case UT: retning = Mottaksretning.UT;
                break;
            case NOTAT: retning = Mottaksretning.NOTAT;
                break;
            default:
                throw new IllegalArgumentException("Journalposttype " + journalposttype.getKode() + " støttes ikke.");
        }
        return retning;
    }
}

File: frontend-api/src/main/kotlin/no/nav/melosys/tjenester/gui/dto/dokumentarkiv/JournalpostInfoDto.kt
package no.nav.melosys.tjenester.gui.dto.dokumentarkiv

import no.nav.melosys.domain.arkiv.ArkivDokument
import no.nav.melosys.domain.arkiv.Journalpost
import no.nav.melosys.domain.arkiv.Journalposttype
import no.nav.melosys.domain.kodeverk.Mottaksretning
import java.time.Instant
import java.util.stream.Collectors

class JournalpostInfoDto(
    @JvmField val journalpostID: String,
    @JvmField val mottattDato: Instant?,
    @JvmField val journalforingDato: Instant,
    @JvmField val mottaksretning: Mottaksretning,
    @JvmField val avsenderEllerMottaker: String,
    @JvmField val hoveddokument: DokumentDto,
    @JvmField val vedlegg: List<DokumentDto>
) {
    fun hentGjeldendeTidspunkt(): Instant? {
        return this.mottattDato ?: this.journalforingDato
    }

    companion object {
        @JvmStatic
        fun av(journalpost: Journalpost): JournalpostInfoDto {
            return JournalpostInfoDto(
                journalpost.journalpostId,
                journalpost.forsendelseMottatt,
                journalpost.forsendelseJournalfoert,
                av(journalpost.journalposttype),
                journalpost.korrespondansepartNavn,
                DokumentDto(
                    journalpost.hoveddokument.dokumentId,
                    journalpost.hoveddokument.tittel, journalpost.hoveddokument.hentLogiskeVedleggTitler()
                ),
                lagVedlegg(journalpost.vedleggListe)
            )
        }

        private fun lagVedlegg(vedlegg: List<ArkivDokument>): List<DokumentDto> {
            return vedlegg.stream()
                .map { v: ArkivDokument -> DokumentDto(v.dokumentId, v.tittel, v.hentLogiskeVedleggTitler()) }
                .collect(Collectors.toList())
        }

        @JvmStatic
        fun av(journalposttype: Journalposttype): Mottaksretning {
            val retning = when (journalposttype) {
                Journalposttype.INN -> Mottaksretning.INN
                Journalposttype.UT -> Mottaksretning.UT
                Journalposttype.NOTAT -> Mottaksretning.NOTAT
                else -> throw IllegalArgumentException("Journalposttype " + journalposttype.kode + " støttes ikke.")
            }
            return retning
        }
    }
}
